### PR TITLE
fix: Indicator filter expression validation

### PR DESCRIPTION
**Pull request by zubaira**

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/function/D2HasValue.java [hunk 1]
```diff
@@ -44,7 +45,9 @@
     @Override
     public Object getDescription( ExprContext ctx, CommonExpressionVisitor visitor )
     {
-        return getProgramArgType( ctx ).getDescription( ctx, visitor );
+        getProgramArgType( ctx ).getDescription( ctx, visitor );
+
+        return DEFAULT_BOOLEAN_VALUE;
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/function/D2HasValue.java:50]: Could you explain the logic here, i.e. we always return the default?

Comment (zubaira) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/function/D2HasValue.java:50]: `getDescription() `function is being used for expression validation and the way it works is all the variables and d2 function are replaced by default values and then the expression is evaluated.

Comment (zubaira) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/function/D2HasValue.java:50]: @jimgrace ?

Comment (jimgrace) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/function/D2HasValue.java:50]: Zubair's fix looks right to me. The `getDescription()` method builds a Map of string substitutions to turn an expression into an expression description, for example replace "#{fbfJHSPpUQD}" with "ANC 1st visit". But what the method *returns* is a dummy value so the overall expression can be syntax-checked.

The ANTLR parsing methods include some type checking. So for example, the first argument to an `if()` function should be boolean, as in `if( test, valueIfTrue, valueIfFalse )`. A program indicator filter should also evaluate to a boolean. The `D2:HasValue()` function should always return a dummy boolean. It was incorrectly trying to return the datatype of the value it was testing.

Comment (larshelge) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/function/D2HasValue.java:50]: Thanks for the feedback.

**Pull request merged**